src/meta/src/manager/sink_coordination/coordinator_worker.rs

```diff
@@ -523,24 +719,117 @@ impl CoordinatorWorker {
-                    Duration::from_secs(5),
-                    || {
-                        warn!(
-                            elapsed = ?start_time.elapsed(),
-                            %sink_id,
-                            "committing"
-                        );
-                    },
-                )
-                .await
-                .map_err(|e| anyhow!(e))?;
-                self.handle_manager
-                    .ack_commit(epoch, commit_requests.handle_ids)?;
+                match &mut coordinator {
+                    SinkCommitCoordinator::SinglePhase(coordinator) => {
+                        let start_time = Instant::now();
+                        run_future_with_periodic_fn(
+                            coordinator.commit(epoch, metadatas, first_add_columns),
+                            Duration::from_secs(5),
+                            || {
+                                warn!(
+                                    elapsed = ?start_time.elapsed(),
+                                    %sink_id,
+                                    "committing"
+                                );
+                            },
+                        )
+                        .await
+                        .map_err(|e| anyhow!(e))?;
+                        self.handle_manager
+                            .ack_commit(epoch, commit_requests.handle_ids)?;
+                    }
+                    SinkCommitCoordinator::TwoPhase(coordinator) => {
+                        let commit_metadata = coordinator
+                            .pre_commit(epoch, metadatas, first_add_columns)
+                            .await?;
+                        persist_pre_commit_metadata(
+                            &db,
+                            sink_id as _,
+                            epoch,
+                            commit_metadata.clone(),
+                        )
+                        .await?;
+                        self.handle_manager
+                            .ack_commit(epoch, commit_requests.handle_ids)?;
+
+                        let two_phase_handler = two_phase_handler.as_mut().expect("should exist");
+                        two_phase_handler.push_new_item(epoch, commit_metadata);
+                    }
+                }
                 prev_commit_epoch = Some(epoch);
             }
         }
     }
+
+    /// Return the log store rewind start epoch if exists.
+    async fn init_state_from_store(
+        &mut self,
+        db: &DatabaseConnection,
+        sink_id: SinkId,
+        subscriber: SinkCommittedEpochSubscriber,
+        coordinator: &mut BoxTwoPhaseCoordinator,
+    ) -> anyhow::Result<(Option<u64>, TwoPhaseCommitHandler)> {
+        let ordered_metadata = list_sink_states_ordered_by_epoch(db, sink_id as _).await?;
+        let mut pending_items = vec![];
+
+        let last_committed_epoch = if ordered_metadata.is_empty() {
```

Comment:
I think the handling logic is the same no matter whether `ordered_metadata.is_empty()` or not, so this branch is unnecessary.